### DISCUSSION TITLE
Wifi Connectivity Issues

### DISCUSSION THREAD

**Discussion by RDunmow** [HomeSpan/HomeSpan#13] [General]
I have been having issues with connecting to WiFi using functionality inside the HomeSpan library in Version 1.1.1. 

The credentials are correct and if I introduce a 1 to 2-second delay from the initial WiFi.begin statement then the wifi will connect.

Not wishing to change the code inside the library, I added this just before the initialisation of the HomeSpan library and the wifi will now connect. 

//Example code, not my real SSID and password
WiFi.begin("{SSID}", "{Password}");
delay(2000);
WiFi.waitForConnectResult();  

It may be down to the TTGO Mini 32 boards that I am using to replace the ESP8266 Wemo Mini Pro that I had been using for my blind control project, but I thought I would share in case this is useful to others.

**Comment by HomeSpan** [comment]
Thanks for this feedback.  I think the issue is related to the different WiFi chips used on different boards.

I re-worked the connection logic so that instead of trying to connect and blocking the rest of the program, HomeSpan now makes just the connection *request* and then returns to the main program.  It then periodically checks back to see if a connection has been made.  If not, it sends a new connection request.

I also have the time-to-wait for a connection before "timing out" grow exponentially.  It starts at 1 second.  If no connection, it sends a new request and waits 2 seconds.  If still no connection, it repeats the process for 4, 8, and then 16 seconds.  If waiting 16 seconds after a connection request still does not work, HomeSpan reports there may a problem with the WiFi, and that it will wait 60 seconds before re-started the connection process.

During this entire time, the device should still be operating, so that if you have pushbuttons that control an appliance, they will still work.

The code has been check into the Development branch.  Please feel free to give it a try.  I will incorporate into the next patch release.

**Comment by RDunmow** [comment]
Thanks for the feedback and updates to the library. I have now tried V1.1.2 and there must be something special with my network configuration. I have done a little work on the WiFi connection routines in the library to make my system reconnect correctly.

On my wifi setup I have multiple TPLINK Omada Access Points which are controlled by a central controller, they all have the same SSID's for the various different WiFi networks that I run.

from lines 328 on the HomeSpan.cpp file I made the following changes

`Serial.print("Attempting connection to: ");
  Serial.print(network.wifiData.ssid);
  Serial.print(". Type 'X <return>' or press Control Button for 3 seconds at any time to terminate search and delete WiFi credentials.");
  
  while(WiFi.status()!=WL_CONNECTED){

    if(nTries++ == 0){
      Serial.print("\nConnecting..");
      WiFi.begin(network.wifiData.ssid,network.wifiData.pwd);
      WiFi.waitForConnectResult();
    }

    if (WiFi.status() == WL_CONNECT_FAILED){
      //Disconnected and try to reconnect
      WiFi.disconnect();
      WiFi.begin(network.wifiData.ssid,network.wifiData.pwd);
      Serial.print("\nRe-Connecting..");
    }

    if(WiFi.status()!=WL_CONNECTED){
`
(After the last statement the rest is the same)

I found that on the first attempt, sometimes the network connection would fail, disconnecting and re-applying the connection details then seems to fix it.

Also on the Web Page that is used to enter credentials, I found a slight problem in that if you have special characters such as an '!' in the password, they will get translated into escape characters in the form which means the connection test will fail as the password is not correct.

Hopefully, some of what I have found is helpful.

Thanks

**Comment by HomeSpan** [comment]
Sorry for the confusion - the code changes I mentioned above are still in the development branch, which I plan on releasing this weekend as v1.1.3.  Among other changes, it implements a form of WiFi.waitForConnectResult(), but does so in the background so that the system does not block while waiting.

I'll also update the web page logic to ensure escaped characters like ! are handled correctly.  Thanks for catching that!

**Reply by RDunmow** [reply]
Cool, sorry I misunderstood, I will wait for 1.1.3 to be released. Just also wanted to say I think the library is brilliant.

**Comment by jakenzhu** [comment]
I carefully followed the instructions for operation, including burning, all other functions are normal, except that there has been no Wi-Fi hot spot “HomeSpan”(LED ,2 times per second)

**Comment by HomeSpan** [comment]
Can you check the following:  From within the Arduino IDE Serial Monitor, type 'A <return>' when connected to HomeSpan.  This should start the hot-spot and you should see the network HomeSpan-Setup show up under Settings->WiFi on your iPhone.  If this does not work, can you verify that the ESP32 can connect to you WiFi network manually by using the 'W' command?

Also, can you send me the log file that appears in the Arduino IDE Serial Monitor when you first start the ESP32 (I assume you are using an ESP32, and not one of the newer ESP32S2 or ESP32C3 chips?
